core/lib/vm_executor/src/oneshot/mod.rs

```diff
@@ -188,39 +189,64 @@ where
-                };
-                vm.push_transaction(transaction);
-                vm.inspect(&mut tracers.into(), InspectExecutionMode::OneTx)
-            });
-            let validation_result = Arc::make_mut(&mut validation_result)
-                .take()
-                .map_or(Ok(()), Err);
-
-            match (exec_result.result, validation_result) {
-                (_, Err(violated_rule)) => Err(ValidationError::ViolatedRule(violated_rule)),
-                (ExecutionResult::Halt { reason }, _) => Err(ValidationError::FailedTx(reason)),
-                _ => Ok(()),
-            }
+            let version = sandbox.env.system.version.into();
+
+            sandbox.execute_in_vm_with_tracer(|vm, transaction| match vm {
+                Vm::Legacy(vm) => {
+                    let (validation_tracer, mut validation_result) =
+                        ValidationTracer::<HistoryDisabled>::new(validation_params, version);
+                    let tracers = vec![validation_tracer.into_tracer_pointer()];
+
+                    vm.push_transaction(transaction);
+                    let exec_result = vm.inspect(&mut tracers.into(), InspectExecutionMode::OneTx);
+
+                    let validation_result = Arc::make_mut(&mut validation_result)
+                        .take()
+                        .map_or(Ok(()), Err);
+
+                    match (exec_result.result, validation_result) {
+                        (_, Err(violated_rule)) => {
+                            Err(ValidationError::ViolatedRule(violated_rule))
+                        }
+                        (ExecutionResult::Halt { reason }, _) => {
+                            Err(ValidationError::FailedTx(reason))
+                        }
+                        _ => Ok(()),
+                    }
+                }
+
+                Vm::Fast(FastVmInstance::Fast(vm)) => {
+                    vm.stop_after_validation();
+                    vm.push_transaction(transaction);
+                    let mut tracer =
+                        vm_fast::validation_tracer::ValidationTracer::new(validation_params);
+                    let result_and_logs = vm.inspect(&mut tracer, InspectExecutionMode::OneTx);
+                    if let Some(violation) = tracer.validation_error() {
+                        return Err(ValidationError::ViolatedRule(violation));
+                    }
+                    match result_and_logs.result {
+                        ExecutionResult::Halt { reason } => Err(ValidationError::FailedTx(reason)),
+                        ExecutionResult::Revert { .. } => {
+                            unreachable!("Revert can only happen at the end of a transaction")
+                        }
+                        ExecutionResult::Success { .. } => Ok(()),
+                    }
+                }
+
+                _ => {
+                    unimplemented!("Shadow validation is not supported.")
+                }
```

Comment:
Why not? AFAICT, `ValidationError`s can be compared (both `Halt`s and `ViolatedValidationRule`s are comparable), so there should be nothing preventing running both VMs with necessary settings and comparing their results.